testsuite/classlibrary/TestCoreUGens.sc

```diff
@@ -645,6 +646,155 @@ TestCoreUGens : UnitTest {
 
 	}
 
+	test_line_startAndEndSampleAreOuputOverDur {
+		var durFrames = 5; // duration for the line to complete its trajectory, in frames
+		var renderFrames = durFrames + 1; // render one extra frame to check that the end value is reached
+		var startVal = 2;
+		var endVal = 20;
+		var tolerance = -100.dbamp;
+		var condvar = CondVar();
+
+		server.bootSync;
+
+		[\ar, \kr].do{ |rate|
+			[Line, XLine].do{ |ugen|
+				var rateScale = switch(rate, \ar, { 1.0 }, \kr, { server.options.blockSize });
+				var fs = server.sampleRate;
+				{
+					ugen.perform(rate, startVal, endVal, (durFrames * rateScale / fs))
+
+				}.loadToFloatArray(renderFrames * rateScale / fs, server,
+					action: { |data|
+
+						this.assertFloatEquals(data.first, startVal,
+							"%.%'s first output sample should be its given start value.".format(ugen, rate), within: tolerance);
+						this.assert(data.last == endVal and: { data[data.size-2] != endVal },
+							"%.% should not reach its end value before the given duration".format(ugen, rate));
+						// Note an exception to the previouse test would be if the duration of the Line is less than a frame duration.
+						// In that (odd) case the "correct" first value is the end value. See GH issue #4279.
+						condvar.signalOne;
+					}
```

Comment:
This whole bit is hard to read. Perhaps define the action function in a variable and format it ...
```
{ ugen.perform(rate, startVal, endVal, (durFrames * rateScale / fs)) }
   .loadToFloatArray(renderFrames * rateScale / fs, server, action: actionFunc);
```

Reply:
I've refactored a bunch of the tests so hopefully it's more compact while still being easy to read.

In this particular case I opted to keep the `action` function explicit an in-place as with other tests. I think it flows more logically for the DSP to be defined first, then the test follows. But hopefully the refactor addresses some of the readability issue you mentioned.

Reply:
Yea this looks much better and depends on less methods and classes than before, testing only what's intended.